Refraction, reflection, and glass

thayerandrews opened the pull request:
- Addresses the remainder of Viktor's refraction feedback by establishing a range for the refraction value. -1 is maximum shrinking of the refraction image, 0.0 is no change, 1.0 is maximum magnification.
- Add CCEffectReflection.
- Add CCEffectGlass which combines both refraction and reflection.
- Address a few low hanging performance issues in CCEffects that were uncovered by Instruments.

Commit:
CCEffectRefraction - Tweak the refraction math a bit

While experimenting with reflection I came up with some modifications to the refraction effect that
yield better results (IMO). I've also come up with a mapping of min and max refraction values that
yield a nice range of visual results (again IMO). 1.0 results in maximum magnification of the environment
image, 0.0 is no change, and -1.0 results in maximum shrinking of the environment image. Please
note that this is based on experimentation and what looks good and I'm not trying to make this an
accurate optical simulation. For instance you might ask:

Q: Why does the refraction value scale the output of the refract() function instead of being used as the
IOR ratio input to refract()?
A: I tried both and I found that the former yields better looking results than the latter.

Q: Where do the min and max refraction values come from?
A: I experimented with a wide range of values and found that these yield pleasing results while
minimizing unexpected behavior.

Commit:
CCEffectRefraction - Condition input refraction values

Map [-1..1] to the min and max values used by GLSL.

Commit:
CCEffectReflection - Initial commit

Commit:
CCEffectReflection - Fill out the guts of the effect

This is mostly copied from CCEffectRefraction.

Commit:
CCEffectReflection - Add fresnel reflectance term to reflection calculation

Commit:
CCEffectReflection - Tweak reflection math

Since we do not currently support cube mapping, using reflect(viewDir, normalDir) to compute
texture coordinates doesn't yield very nice looking results. It does create some nice distortion
but (to my eyes at least) it doesn't read as reflection and looks more like refraction. By experimentation
I've found the following looks better:

texCoordOffset = normalDir * (1.0 - dot(viewDir, normalDir))

Also, clamping the offset texture coordinates to [0..1] doesn't look as good for reflection as it does for
refraction. Instead, feed them through a cosine function so out of range values reflect.

Commit:
CCEffectRefraction - Update comments - No functional change

Commit:
CCEffectRefraction - Remove enableFresnel property since it can be disabled with the appropriate inputs.

Commit:
CCEffectsTest - Add reflection test

Commit:
CCEffectsTest - Tweak the reflection test

Commit:
CCEffectGlass - Initial commit

Commit:
CCEffectGlass - WIP

Commit:
CCEffectGlass - Flesh out all the refraction and reflection math

A bunch of this should be refactored into a common location that is shared by
CCEffectRefraction, CCEffectReflection, and CCEffectGlass.

Commit:
CCEffectsTest - Add a glass test.

Commit:
CCEffectReflection - Fix effect function name

Commit:
Test data for CCEffectsTest and the associated license file.

Commit:
CCEffectsTest - Update reflection, refraction, and glass test data.

Commit:
CCEffectReflection / Glass - Fake reflection math tweaks.

Increase the power that (1 - nDotV) is raised to and then scale the result
back into the range [0..1]. Since nDotV is proportional to the cosine of the
angle between N and V, raising it to a power > 1 has the effect of flattening
out the center of the curve which looks better to me when applied to a
spherical normal map.

I should have been scaling (1-nDotV) all along since it was ending up in the
range [0..2] which resulted in excessive compression of the environment map
image at the silhouette edges of the normal map. This only gets worse when
raising to a power so the scale is even more important here.

Commit:
Add CCEffectUtils - A location for shared effect functions

Commit:
CCEffects - Refactor reflection, refraction, and glass effects to use shared util functions

Commit:
CCEffectsTest - Adjust refraction values

And fix the alpha channel in the torus normal map.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Xcode project - Move CCEffectGlass within the Effects group

Commit:
CCEffects - Add debug label to CCEffectRenderPass

While profiling the effects code I noticed that construction of each render pass's debug
label via [NSString stringWithFormat] was pretty high on the list of expensive code paths.
Instead of constructing this dynamically during rendering it's now created when the render
passes are created.

This results in a 14% performance boost in the effect performance test with the saturation
effect.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
CCEffect - Copy uniforms when creating the current render state

This marks the render state as immutable and prevents a copy of the entire render
state object when it is enqueued by the renderer.

The next change is to stop recreating render state objects with every draw.

Commit:
CCEffects - Fix debug label creation

It doesn't need to be in the pass's begin block.

Commit:
CCEffectSaturation - Store conditioned saturation as NSNumber instead of float

Which most importantly allows us to not create a new NSNumber from the conditioned float
on every draw. It's a small performance win but it was easy to achieve so it's worthwhile.

Commit:
CCEffectRenderer - Move pushGroup/popGroup into debug-only macros

They're only used in this context for debugging (by identifying which GL commands are
connected to which effect passes) and they have a non-negligible impact on performance
so I'm making them optional and disabled by default.

Commit:
CCEffectHue - Store conditioned hue / matrix as NSNumber / NSValue

To avoid recreating these on every draw.

Commit:
CCEffectBrightness / CCEffectContrast - Store conditioned inputs as NSNumber

And don't create NSNumbers on every draw.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

slembcke commented on cocos2d/CCEffectRenderer.m:
```diff
@@ -19,6 +19,16 @@
+#define CCEFFECT_DEBUG_PUSH_GROUP(R) [(R) pushGroup]
+#define CCEFFECT_DEBUG_POP_GROUP(R, L) [(R) popGroupWithDebugLabel:(L) globalSortOrder:0]
+#else
+#define CCEFFECT_DEBUG_PUSH_GROUP(R)
```
Skipping the push/pop of the group for debug purposes isn't quite right. The blur effects break for sure. Did you just want to skip the debug label generation?

thayerandrews replied:
Okay. The way I remembered it was that I introduced the push/pop to aid in debugging the corresponding GL calls but maybe the push/pop was already there. First I fixed debug label creation (doing it ahead of time and storing it in CCEffectRenderPass instead of doing it repeatedly on the fly) but then I still saw that push/pop was fairly expensive so I switched to the macros.

Thanks for pointing out the issue. I'll dig further into it.

slembcke commented:
Looks neat. Except for the push/pop bug it looks fine.

Don't worry about it for this request (yours are okay), but do be a little careful with testing asset file sizes. A few months ago we rewrote the repository history to slim it down so it wasn't such a pain to use as a submodule. One of the smaller, but significant things removed was ~50MB of old unused testing assets.

Commit:
CCEffectRenderer - Fix bug introduced by removal of pushGroup / popGroup

Scott pointed out that this change had broken the blur effects so I'm rolling it back.

thayerandrews commented:
The pushGroup / popGroup change is rolled back. Thank you for catching that.

slembcke commented:
Push/pop can create a lot of array garbage if there are a lot of calls to them. It wasn't a priority in the past, but this is the second that pooling the queues has come up, so I'll bump up the priority. ;)

jtwigg commented:
@slembcke If this ones ready to go, can we get it pull'd? I'm working on the SB UI and waiting for these changes.

Cheers.
